### PR TITLE
Add support for IVsLanguageBlock.GetCurrentBlock

### PR DESCRIPTION
**Pull request by Pilchie**
Fixes internal TFS bug 1043580.

### CODE REVIEW

**Comment by Pilchie** [comment]
@jasonmalinowski @dpoeschl @balajikris @basoundr @brettfo @rchande any comments?

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            VsTextSpan[] ptsBlockSpan,
+            out string pbstrDescription,
+            out int pfBlockAvailable)
+        {
+            var foundBlock = false;
+            string description = null;
+            var span = default(TextSpan);
+
+            var snapshot = this.EditorAdaptersFactoryService.GetDataBuffer(pTextLines).CurrentSnapshot;
+            var position = snapshot.GetPosition(iCurrentLine, iCurrentChar);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: Is there any expected error handling if this line or the line above fail?

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    internal static class VsLanguageBlock
+    {
+        public static bool GetCurrentBlock(
+            ITextSnapshot snapshot,
+            int position,
+            CancellationToken cancellationToken,
+            ref string description,
+            ref TextSpan span)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: What happens if it's null?

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            var position = snapshot.GetPosition(iCurrentLine, iCurrentChar);
+
+            var waitIndicator = this.Package.ComponentModel.GetService<IWaitIndicator>();
+            waitIndicator.Wait(
+                ServicesVSResources.CurrentBlock,
+                ServicesVSResources.DeterminingCurrentBlock,
+                allowCancel: true,
+                action: context =>
+                {
+                    foundBlock = VsLanguageBlock.GetCurrentBlock(snapshot, position, context.CancellationToken, ref description, ref span);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: What is the expected contract if the user cancels? Do we return an HRESULT here or is "I didn't find anything" a better answer?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 4]
```diff
@@ -721,9 +721,12 @@ public SyntaxNode GetContainingMemberDeclaration(SyntaxNode root, int position)
             var node = root.FindToken(position).Parent;
             while (node != null)
             {
-                if (node is MemberDeclarationSyntax)
+                if (useFullSpan || node.Span.Contains(position))
                 {
-                    return node;
+                    if (node is MemberDeclarationSyntax)
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Why the change here? What about compound things like accessors or something? What should be returned there?

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb [hunk 5]
```diff
@@ -593,25 +593,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                    If TypeOf node Is EventStatementSyntax AndAlso Not TypeOf node.Parent Is EventBlockSyntax Then
+                        Return node
+                    End If
+
+                    If TypeOf node Is PropertyBlockSyntax OrElse
+                       TypeOf node Is TypeBlockSyntax OrElse
+                       TypeOf node Is EnumBlockSyntax OrElse
+                       TypeOf node Is NamespaceBlockSyntax OrElse
+                       TypeOf node Is EventBlockSyntax OrElse
+                       TypeOf node Is FieldDeclarationSyntax Then
```
Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: Declare Sub? And where do delegates land?

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb [hunk 6]
```diff
@@ -593,25 +593,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
-                   TypeOf node Is FieldDeclarationSyntax Then
-                    Return node
+                    If TypeOf node Is EventStatementSyntax AndAlso Not TypeOf node.Parent Is EventBlockSyntax Then
+                        Return node
+                    End If
+
+                    If TypeOf node Is PropertyBlockSyntax OrElse
+                       TypeOf node Is TypeBlockSyntax OrElse
+                       TypeOf node Is EnumBlockSyntax OrElse
+                       TypeOf node Is NamespaceBlockSyntax OrElse
```
Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: VB returns namespaces but C# doesn't?

Comment (Pilchie) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: `NamespaceDeclarationSyntax` : `MemberDeclarationSyntax` in C#.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs (hunk 4 above)
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Don't want to find the node if position is in it's trivia, instead, we want to keep walking up to the parent node.

Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Accessors should (and do) return their property containing property.

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb (hunk 5 above)
Comment (Pilchie) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: Note that I didn't write this function, just indented it try &w=0 in the diff :smile:.

However, since they don't have bodies, I'm totally find with saying they aren't a "Block".  AFAIK, this interface is only used to populate the current block scope in the Find in Files dialog, so I'm not *terribly* about things like this.

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs (hunk 1 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: Handled.

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs (hunk 2 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: Good catch.

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs (hunk 3 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: I tested adding a `throw new OperationCanceledException();` here, and the find dialog handled it the same as returning nothing found (doesn't show the current block scope as an option).  I'm inclined to leave the code as is.

Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.IVsLanguageBlock.cs]: Fine by me. My min-bar was just "this was thought about". :-)

**Comment by Pilchie** [comment]
So, @jasonmalinowski - is that a :+1:?

Anyone else?

**Comment by dpoeschl** [comment]
LGTM

**Comment by jasonmalinowski** [comment]
:+1:
